src/sql/workbench/services/diagnostics/browser/errorDiagnosticsService.ts

```diff
@@ -0,0 +1,42 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IErrorDiagnosticsService } from 'sql/workbench/services/diagnostics/common/errorDiagnosticsService';
+import * as azdata from 'azdata';
+
+export class ErrorDiagnosticsService implements IErrorDiagnosticsService {
+
+	_serviceBrand: undefined;
+	private _providers: { [handle: string]: azdata.diagnostics.ErrorDiagnostics; } = Object.create(null);
+
+	constructor(
+	) { }
+
+	public async checkConnectionError(errorCode: number, errorMessage: string, providerId: string, connection: azdata.connection.ConnectionProfile, options: azdata.IConnectionCompletionOptions): Promise<boolean> {
+		let result = false;
+		let provider = this._providers[providerId]
+		if (provider) {
+			result = await provider.handleConnectionError(errorCode, errorMessage, connection, options);
+		}
+		return result;
+	}
+
+	/**
+	 * Register a diagnostics object for a provider
+	 * @param providerId the id of the provider to register.
+	 * @param diagnostics the actual diagnostics provider object to register under the id.
+	 */
+	public registerDiagnosticsProvider(providerId: string, diagnostics: azdata.diagnostics.ErrorDiagnostics): void {
+		this._providers[providerId] = diagnostics;
```

Comment:
So right now you only allow a single provider to be registered per connection provider. It seems reasonable that if you're going to allow ANY extension to register itself for a given connection provider that we should allow multiple extensions to register a handler for the same connection provider.

Although of course this brings up the question - do we even want to allow that in the first place? Another option would be to actually have this functionality be an extension on the connection provider instead so that ONLY the actual extension that manages the connection is able to handle errors for it. (this goes back to my whole thing about this not seeming to be something that belongs in the "diagnostics" namespace in the first place and rather is an extension on the connection provider stuff itself)


@cheenamalhotra @alanrenmsft

Reply:
I'm going with the design that was agreed upon earlier which was to have a single provider for each area, but if you want to have any extension register their own for the provider, then that means I'll have to include an array and await them all, which I can do.

Reply:
This is designed the same way as other SQL workbench services:
-  accountManagementService
- executionPlanService
They all work with single provider registration, and so we've kept the same pattern.

DiagnosticsService is also not designed for only connection provider, so it shouldn't be included with connection provider, it needs to be generic enough to be able to handle errors for different areas.

If there's an ask in future to allow multiple registrations, we can change it internally later, but I don't think we need to change anything as of now.

Reply:
I know that we haven't always been great about how provider registering is handled in the past so there's very likely issues (which doesn't change what I brought up - just because it may have been done in a certain way in the past doesn't mean we need to repeat the same mistakes here)

And to be clear I never said anything about the whole diagnostics provider stuff you were talking about adding in the future being included in the connection provider. This is just the connection error handling provider (which I still think belongs in the connection namespace and not here)

Anyways, if you're going to go forward with the current design of only allowing one provider to be registered then you should at least update the docs to make that explicitly clear so people aren't confused when they run into that.

Reply:
Well then, I'd say we need to fix existing designs first or atleast document/comment somewhere about mistakes/patterns to avoid. Because for new developers these are references of a "working" concept. Same goes for all the old code writing practices around async promises. It's not clear looking at the code whether this is a "practice to follow" or just an old code. Because its not unnatural to keep using old design patterns in big applications due to any limitations.

Reply:
I am very much not against fixing/improving existing behavior or adding more documentation 😄 Not really a blocker for trying to make it better here though - change has to start somewhere after all.